auraed/src/vms/virtual_machine.rs

```diff
@@ -0,0 +1,382 @@
+        DebugConsoleConfig, HotplugMethod, MemoryConfig, PayloadConfig,
+        RngConfig, VhostMode, DEFAULT_DISK_NUM_QUEUES, DEFAULT_DISK_QUEUE_SIZE,
+        DEFAULT_MAX_PHYS_BITS, DEFAULT_NET_NUM_QUEUES, DEFAULT_NET_QUEUE_SIZE,
+    },
+    vm::VmState,
+};
+
+use crate::vms::manager::Manager;
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct VmID(String);
+
+impl VmID {
+    pub fn new(id: impl Into<String>) -> Self {
+        Self(id.into())
+    }
+}
+
+impl Display for VmID {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct VmSpec {
+    pub memory_size: u32,
+    pub vcpu_count: u32,
+    pub kernel_image_path: PathBuf,
+    pub kernel_args: Vec<String>,
+    pub mounts: Vec<MountSpec>,
+    pub net: Vec<NetSpec>,
+}
+
+impl From<VmSpec> for vmm::vm_config::VmConfig {
+    fn from(spec: VmSpec) -> Self {
+        vmm::vm_config::VmConfig {
+            cpus: CpusConfig {
+                boot_vcpus: spec.vcpu_count as u8,
+                max_vcpus: spec.vcpu_count as u8,
+                topology: None,
+                kvm_hyperv: false,
+                max_phys_bits: DEFAULT_MAX_PHYS_BITS,
+                affinity: None,
+                features: CpuFeatures::default(),
+            },
+            memory: MemoryConfig {
+                size: (spec.memory_size << 20) as u64,
+                mergeable: false,
+                hotplug_method: HotplugMethod::default(),
+                hotplug_size: None,
+                hotplugged_size: None,
+                shared: false,
+                hugepages: false,
+                hugepage_size: None,
+                prefault: false,
+                zones: None,
+                thp: false,
+            },
+            payload: Some(PayloadConfig {
+                firmware: None,
+                kernel: Some(spec.kernel_image_path),
+                cmdline: Some(spec.kernel_args.join(" ")),
+                initramfs: None,
+            }),
+            rate_limit_groups: None,
+            disks: Some(spec.mounts.into_iter().map(Into::into).collect()),
+            net: Some(spec.net.into_iter().map(Into::into).collect()),
+            rng: RngConfig::default(),
+            balloon: None,
+            fs: None,
+            pmem: None,
+            serial: default_serial(),
+            console: default_console(),
+            debug_console: DebugConsoleConfig::default(),
+            devices: None,
+            user_devices: None,
+            vdpa: None,
+            vsock: None,
+            pvpanic: false,
+            iommu: false,
+            sgx_epc: None,
+            numa: None,
+            watchdog: false,
+            pci_segments: None,
+            platform: None,
+            tpm: None,
+            preserved_fds: None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NetSpec {
+    pub tap: Option<String>,
+    pub ip: Ipv4Addr,
```

Comment:
can we do Ipv6 by default? or does that give us some compatibility issues?

Reply:
I'm just letting Cloud Hypervisor set up the networking here, which seems to automatically set an IPv4 address by as well as an IPv6 link-local by default. I think we would just need to set up the tap device manually to allow this to be configurable.

```
tap0: flags=67<UP,BROADCAST,RUNNING>  mtu 1500
        inet 192.168.249.1  netmask 255.255.255.0  broadcast 192.168.249.255
        inet6 fe80::68af:9eff:fed6:8823  prefixlen 64  scopeid 0x20<link>
        ether 06:66:42:a8:3f:e1  txqueuelen 1000  (Ethernet)
        RX packets 9  bytes 822 (822.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 21  bytes 2182 (2.1 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

Reply:
fair enough. thanks for confirming